client/src/main/java/io/pravega/client/stream/impl/StreamSegments.java

```diff
@@ -71,13 +75,20 @@ public StreamSegments withReplacementRange(StreamSegmentsWithPredecessors replac
         verifyReplacementRange(replacementRanges);
         NavigableMap<Double, Segment> result = new TreeMap<>();
         Map<Long, List<SegmentWithRange>> replacedRanges = replacementRanges.getReplacementRanges();
-        for (Entry<Double, Segment> exitingSegment : segments.entrySet()) {
-            List<SegmentWithRange> replacements = replacedRanges.get(exitingSegment.getValue().getSegmentId());
+        for (Entry<Double, Segment> existingSegment : segments.entrySet()) {
+            List<SegmentWithRange> replacements = replacedRanges.get(existingSegment.getValue().getSegmentId());
             if (replacements == null || replacements.isEmpty()) {
-                result.put(exitingSegment.getKey(), exitingSegment.getValue());
+                result.put(existingSegment.getKey(), existingSegment.getValue());
             } else {
                 for (SegmentWithRange replacement : replacements) {
-                    result.put(replacement.getHigh(), replacement.getSegment());
+                    /*
+                      In case of a segment merge the replacement segment's high key can be greater than or equal to the
+                      existing segment high key. In this scenario the existing segment's high key should be used.
+
```

Comment:
Won't this leave a hole in the keyspace? Will help if you could add a sample example.

Reply:
Refactored the implementation based on the discussions in the PR; added junits samples for the same covering different scenarios.